presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java

```diff
@@ -282,4 +277,56 @@ private boolean inPredicateMatchesApply(ApplyNode node, Symbol value, Symbol val
             return ImmutableMap.of(originalInPredicate, semiJoinSymbol.get().toSymbolReference());
         }
     }
+
+    /**
+     * Calls {@link ApplyNodeRewriter#rewriteApply(ApplyNode)}} for Apply which produces given Expression
+     */
+    public abstract static class ApplyNodeRewriter
+            extends SimplePlanRewriter<Void>
+    {
+        protected SymbolReference symbolReference;
+
+        public ApplyNodeRewriter(SymbolReference symbolReference)
+        {
+            this.symbolReference = requireNonNull(symbolReference, "symbolReference is null");
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            Expression expression = replaceExpression(
+                    symbolReference,
+                    mapAssignmentSymbolsToExpression(node.getAssignments()));
+
+            if (expression instanceof SymbolReference) {
+                symbolReference = (SymbolReference) expression;
+                return context.defaultRewrite(node, context.get());
+            }
+            else {
+                return node;
+            }
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<Void> context)
+        {
+            if (inPredicateMatchesApply(node)) {
+                return rewriteApply(node);
+            }
+            return context.defaultRewrite(node, context.get());
+        }
+
+        protected abstract PlanNode rewriteApply(ApplyNode node);
+
+        private boolean inPredicateMatchesApply(ApplyNode node)
```

Comment:
Why is does the name of this method refer to "inPredicate"? There doesn't seen to be anything specific to that kind of predicate in its body.